### DISCUSSION TITLE
Non-Sharded Functions

### DISCUSSION THREAD

**Discussion by Legends-of-NGDB** [discordjs/discord.js#10573] [Q&A]
Hello again. I think I have discovered in my other discussion why I'm sometimes hitting the rate limit for broadcastEval, but not totally.

```js
mongoose.connect(mongoURL || '', { }).then(() => {
 
  setTimeout(() => {
    setInterval(() => { checkUpdates(client) }, 130_000);
    
   // setTimeout(() => {
     // setInterval(() => { twitch(client); }, 80_000);
      
     // setTimeout(() => {
      //  setInterval(() => { antibot(client); }, 30_000);
        
        setTimeout(() => {
          setInterval(() => { checkGiveaway(client); }, 5000);
          
          setTimeout(() => {
            setInterval(() => { checkAIChannels() }, 10_000);
          }, 20000);
        }, 20000);
      //}, 20000);
   // }, 20000);
  }, 20000);

})
```
These are function that I call for broadcastEval() and they are happening in my bot.js so __every__ shard is calling them. This is very unnecessary as the function to run on its own shard. I haven't reached 2500 guilds yet, but is there a work around without a discord.js error about sharding? I was thinking something like this:

```js
const { Client, GatewayIntentBits, Partials, ShardingManager } = require('discord.js');
    require('dotenv').config();

//this is just another client, don't concern yourself with this one
    const scrapper = new Client({
        intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMembers,
        GatewayIntentBits.GuildMessages,
        ],
        partials: [Partials.User, Partials.GuildMember],
      });

    scrapper.login(process.env.scrapper)

const manager = new ShardingManager('./bot.js', { token: process.env.token, shards: 'auto' });

manager.on('shardCreate', shard => console.log(`Launched shard ${shard.id}`));

manager.spawn();

const ngdb = new Client({
  intents: [
  //GatewayIntentBits.GuildPresences,
  GatewayIntentBits.Guilds,
  GatewayIntentBits.GuildMembers,
  GatewayIntentBits.GuildMessages,
  GatewayIntentBits.GuildMessageReactions,
  GatewayIntentBits.MessageContent,
  ],
  partials: [Partials.Channel, Partials.User, Partials.Message, Partials.Reaction, Partials.GuildMember],
});

ngdb.login(process.env.token)

//import the function here and pass client as "ngdb"

module.exports = { scrapper, manager, ndgb}

```

Would this example work?

**Comment by Legends-of-NGDB** [comment]
My reaction event also is running on shards when I'm using broadcastEval. This probably is why I'm encountering it


```js
const ngdb = require(`./../@ngdb`);
const reactions = require(`./../Schemas/react.js`);

module.exports = async (client, reaction, user) => {
console.log(reaction.message.guild)
if (user.bot || !reaction.message.guildId) return;

ngdb.AddBuilder(client, reaction, user);

return reactionRole(client, reaction, user);

}

async function reactionRole(client, reaction, user) {
try {

    let emoji;
    if (reaction.emoji.id) {
        // Custom emoji
        emoji = `<:${reaction.emoji.name}:${reaction.emoji.id}>`;
    } else {
        // Unicode emoji
        emoji = reaction.emoji.name;
    }

    const data = await reactions.findOne({ Guild: reaction.message.guildId, Message: reaction.message.id, Emoji: emoji });

    if (!data) return;
    

    const guildEval = reaction.message.guildId;

    const shardId = Discord.ShardClientUtil.shardIdForGuildId(guildEval, client.shard.count);
  
    client.shard.broadcastEval(
      async (c, { guildEval }) => {
        if (c.guilds.cache.has(guildEval)) {
  
const guild = await c.guilds.fetch(guildEval)
const member = await guild.members.cache.get(user.id);

    try {
        await member.roles.add(data.Role);
    } catch (e) {
    return;
    }

         }
        },
        {
            context: { guildEval },
            shard: shardId
          })

  
} catch (e) {

}
}
```

**Comment by RlxChap2** [comment]
I noticed that your current approach with sharding and `broadcastEval` in Discord.js could lead to issues like hitting rate limits and unnecessary duplication of logic. Here's a more efficient way to handle things:

Sharding is essential when running bots on multiple servers, but it can lead to issues if not handled properly. In your case, calling `broadcastEval` across **all shards** for the same event can cause multiple executions of the same logic, increasing the likelihood of rate limits and inefficiencies.

- **Multiple Clients**: You're using separate clients (`scrapper`, `ngdb`) for different tasks, which adds unnecessary complexity and resource consumption. Instead of multiple clients, it’s better to manage everything through a single sharded client.
  
- **Broadcasting to All Shards**: You're broadcasting events using `client.shard.broadcastEval` across all shards. However, many of these tasks (like adding roles in a guild) likely only need to be done on the shard responsible for that guild. Broadcasting across all shards can quickly hit rate limits or create redundant operations.

Instead of broadcasting to all shards, you should target the shard responsible for a particular guild. This can be done using `Discord.ShardClientUtil.shardIdForGuildId()`, which allows you to find the correct shard ID based on the guild’s ID. This avoids unnecessary execution across all shards.

Example:
```javascript
const shardId = Discord.ShardClientUtil.shardIdForGuildId(reaction.message.guildId, client.shard.count);

if (client.shard.id === shardId) {
    // Execute task on the correct shard only
    await handleReactionRole(client, reaction, user, data);
}
```

You’re using multiple `setInterval` and `setTimeout` calls in different places, which can cause performance issues and redundant operations when they run on every shard. Instead, let a single shard handle the scheduling of recurring tasks like `checkUpdates`, `checkGiveaway`, and so on. You can then broadcast the results to other shards as needed.

For example:
- Have one central shard handle the timers and delegate tasks to other shards when necessary.

If a function should only run on one shard, ensure it’s not broadcasted to every shard. In your `reactionRole` function, you can target the correct shard based on the guild ID to ensure only one shard processes the reaction event.

For instance:
```javascript
async function reactionRole(client, reaction, user) {
    try {
        const emoji = reaction.emoji.id ? `<:${reaction.emoji.name}:${reaction.emoji.id}>` : reaction.emoji.name;
        const data = await reactions.findOne({ Guild: reaction.message.guildId, Message: reaction.message.id, Emoji: emoji });

        if (!data) return;

        // Determine the correct shard
        const shardId = Discord.ShardClientUtil.shardIdForGuildId(reaction.message.guildId, client.shard.count);

        if (client.shard.id === shardId) {
            await handleReactionRole(client, reaction, user, data);
        }
    } catch (e) {
        console.error(e);
    }
}
```

This ensures that you don’t broadcast tasks across all shards, but only to the shard responsible for that particular guild.

Instead of creating multiple clients (`scrapper`, `ngdb`), you should use a **single ShardingManager** to manage all your shards efficiently. You can define all functionality within a single bot process, which simplifies your setup and reduces overhead.

```javascript
const { ShardingManager } = require('discord.js');
require('dotenv').config();

const manager = new ShardingManager('./bot.js', {
    token: process.env.token,
    shards: 'auto'
});

manager.on('shardCreate', shard => console.log(`Launched shard ${shard.id}`));

manager.spawn();
```

In `bot.js`, you can then use the `client` object managed by the `ShardingManager` to execute all tasks across your shards.

- **Rate Limiting**: Be mindful of the API rate limits. Broadcasting the same task to all shards can easily exceed the limits, especially when the bot has a large number of guilds. Always target only the necessary shards to prevent unnecessary API calls.

- **Centralized Task Handling**: Instead of having every shard handle timers like `checkUpdates`, consider centralizing these tasks. Have one shard (or process) manage recurring functions and only pass data to other shards when necessary.

By ensuring that tasks are only executed on the relevant shard and avoiding the unnecessary duplication of code across shards, you can significantly improve your bot’s performance. This setup not only reduces the risk of hitting rate limits but also makes the bot more efficient overall.

I hope this helps optimize your bot and avoid sharding-related issues! Feel free to reach out if you need further clarification.

**Reply by Qjuh** [reply]
Don’t post AI responses. Like most AI responses it‘s long, convoluted and - most importantly - wrong.
